common/workunit/workunit.cpp

```diff
@@ -874,6 +884,106 @@ class CDaliWuGraphStats : public CWuGraphStats
     Owned<IRemoteConnection> conn;
 };
 
+CWorkUnitWatcher::CWorkUnitWatcher(IWorkUnitSubscriber *_subscriber, WUSubscribeOptions flags, const char *wuid) : subscriber(_subscriber)
+{
+    abort = 0;
+    state = 0;
+    action = 0;
+    assertex((flags & ~SubscribeOptionAbort) == 0);
+    if (flags & SubscribeOptionAbort)
+    {
+        VStringBuffer xpath("/WorkUnitAborts/%s", wuid);
+        abort = querySDS().subscribe(xpath.str(), *this);
+    }
+}
+CWorkUnitWatcher::~CWorkUnitWatcher()
+{
+    assertex(abort==0 && state==0 && action==0);
+}
+
+void CWorkUnitWatcher::unsubscribe()
+{
+    CriticalBlock b(crit);
+    if (abort)
+        querySDS().unsubscribe(abort);
+    if (state)
+        querySDS().unsubscribe(state);
+    if (action)
+        querySDS().unsubscribe(action);
+    abort = 0;
+    state = 0;
+    action = 0;
```

Comment:
The assert will ensure that....

This code was structured to require explicit unsubscribe rather than relying on destruction - related to destructor order, because typically by the time the destructor is called you are already in trouble (since the subscriber is most likely destroyed - or at least partially destroyed - by then, as usually the subscriber owns the watcher).

I didn't change this paradigm, but I think that's why it is there.